app/models/manageiq/providers/redhat/infra_manager/ovirt_services/strategies/v3.rb

```diff
@@ -17,5 +25,302 @@ def cluster_name_href(href)
+        rhevm_vm.attributes.fetch_path(:status, :state) == "image_locked"
+      end
+    end
+
+    def populate_phase_context(phase_context, vm)
+      phase_context[:new_vm_ems_ref] = ManageIQ::Providers::Redhat::InfraManager.make_ems_ref(vm[:href])
+      phase_context[:clone_task_ref] = vm.creation_status_link
+    end
+
+    def nics_for_(vm)
+      vm.with_provider_object do |rhevm_vm|
+        rhevm_vm.nics.collect { |n| NicsDecorator.new(n) }
+      end
+    end
+
+    def cluster_find_network_by_name(href, network_name)
+      ext_management_system.with_provider_connection do |rhevm|
+        Ovirt::Cluster.find_by_href(rhevm, href).try(:find_network_by_name, network_name)
+      end
+    end
+
+    def configure_vnic(args)
+      vm = args[:vm]
+      mac_addr = args[:mac_addr]
+      network = args[:network]
+      nic_name = args[:nic_name]
+      interface = args[:interface]
+      vnic = args[:vnic]
+      logger = args[:logger]
+
+      options = {
+        :name        => nic_name,
+        :interface   => interface,
+        :network_id  => network[:id],
+        :mac_address => mac_addr,
+      }.delete_blanks
+
+      logger.info("with options: <#{options.inspect}>")
+
+      if vnic.nil?
+        vm.with_provider_object do |rhevm_vm|
+          rhevm_vm.create_nic(options)
+        end
+      else
+        vnic.apply_options!(options)
+      end
+    end
+
+    def powered_off_in_provider?(vm)
+      vm.with_provider_object(&:status)[:state] == "down"
+    end
+
+    def powered_on_in_provider?(vm)
+      vm.with_provider_object(&:status)[:state] == "up"
+    end
+
+    def vm_boot_from_cdrom(operation, name)
+      operation.get_provider_destination.boot_from_cdrom(name)
+    rescue Ovirt::VmNotReadyToBoot
+      raise OvirtServices::VmNotReadyToBoot
+    end
+
+    def vm_boot_from_network(operation)
+      operation.get_provider_destination.boot_from_network
+    rescue Ovirt::VmNotReadyToBoot
+      raise OvirtServices::VmNotReadyToBoot
+    end
+
+    def get_template_proxy(template, connection)
+      connection ||= ext_management_system.connect
+      template_proxy = connection.get_resource_by_ems_ref(template.ems_ref)
+      GeneralUpdateMethodNamesDecorator.new(template_proxy)
+    end
+
+    def get_vm_proxy(vm, connection)
+      connection ||= ext_management_system.connect
+      vm_proxy = connection.get_resource_by_ems_ref(vm.ems_ref)
+      GeneralUpdateMethodNamesDecorator.new(vm_proxy)
+    end
+
+    def collect_disks_by_hrefs(disks)
+      vm_disks = []
+
+      ext_management_system.try(:with_provider_connection) do |rhevm|
+        disks.each do |disk|
+          begin
+            vm_disks << Ovirt::Disk.find_by_href(rhevm, disk)
+          rescue Ovirt::MissingResourceError
+            nil
+          end
+        end
+      end
+      vm_disks
+    end
+
+    def shutdown_guest(operation)
+      operation.with_provider_object(&:shutdown)
+    rescue Ovirt::VmIsNotRunning
```

Comment:
This logic was not changed from before. I found many things I do not like but kept the same logic because this is out of the scope of this PR.